react-js/react-components-and-hierarchies.md

```diff
@@ -0,0 +1,438 @@
+*Example*
+
+Parent.js
+```
+
+class Parent extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      fromChild: "empty"
+    };
+  }
+
+  update = data => {
+    this.setState({ fromChild: data });
+  };
+
+  render() {
+    return (
+      <div className="App">
+        <h1>Parent Component </h1>
+        <p>
+          data coming from child<b>: {this.state.fromChild} </b>
+        </p>
+        <Child update={this.update} />
+      </div>
+    );
+  }
+}
+
+
+```
+Child.js
+```
+class Child extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      child: "Data received"
+    };
+  }
+
+  update = data => {
+    this.setState({ fromChild: data });
+  };
+
+  render() {
+    return (
+      <div className="App">
+        <h1>Child Component </h1>
+        <button onClick={() => this.props.update(this.state.child)}>
+          Click from Child
+        </button>
+      </div>
+    );
+  }
+}
+
+export default Child;
+```
+[Run code](https://codesandbox.io/s/pass-data-from-child-to-parent-6tl0e?file=/src/Child.js:43-493)
+
+
+
+### What are props.children
+1. Read [A quick intro to React’s props.children](https://medium.com/better-programming/passing-data-to-props-children-in-react-5399baea0356)
+2. Read [React This Props Children](https://learn.co/lessons/react-this-props-children)
+3. Watch [React Tutorial 13: props.children](https://www.youtube.com/watch?v=Sq0FoUPxj_c)
+
+The ability for components to receive and render child elements is one of the most important feature of react. This makes it really easy to create reusable components. All we need to do is to  wrap props.children with some markup or behavior props.children does  is used to display whatever we include between the opening and closing tags when invoking a component.
+
+*Example*
+```
+class MyComponent extends React.Component {
+  render (){ 
+    return (
+      <div>
+        <h1>MyComponent JSX here</h1>
+        {this.props.children}
+      </div>
+    );
+  }
+}
+
+
+
+ReactDOM.render(
+  <MyComponent>
+    <p>understanding prop.children </p>
+  </MyComponent>,
+  document.getElementById('app')
+);
+```
+
+
+Whenever the MyComponent is invoked {props.children} will also be displayed and this is just a reference to what is between the opening and closing tags of the component.
+
+[Run code](https://codepen.io/annu12340/pen/wvKowmj?editors=1010)
+
+Instead of invoking the component with a self-closing tag < MyComponent /> if you invoke it will full opening and closing tags <MyComponent> </MyComponent> you can then place more code between it.
+ 
+This de-couples the <MyComponent> component from its content and makes it more reusable.
+
+The possible usage for {props.children} are:
+
+- Grouping unknown number of similar elements into a parent element.
+- You don’t know elements ahead of the time.
+- The nested structure that needs a wrapper.
+
+
+### Using 3p components like reactstrap
```

Comment:
Please use "third-party" spelled out instead of "3p" - I was just typing shorthand in the issue.